### PR TITLE
feat(DPGO-19, 21): email sender, password restore story

### PR DESCRIPTION
**Pull request by glbter**
- gmail oauth email sender with account dp210go
- password restore feature
- change password feature

### CODE REVIEW

File: internal/server/http/user/handlers.go [hunk 1]
```diff
@@ -26,18 +26,25 @@ type UsersUsecases interface {
...
+
 const idKey = "id"
 const tokenTime = time.Minute * 15
 
 // Handlers represent a user handlers.
 type Handlers struct {
-	usecases UsersUsecases
-	logger   *zap.Logger
+	userCases UsersUsecases
+	paswCases PasswordUsecases
```
Comment (igkostyuk) [inline internal/server/http/user/handlers.go]: why do we need  it here ? can we have separate handlers ?

File: internal/repository/postgres/user/repo.go [hunk 2]
```diff
@@ -114,6 +114,16 @@ func (r *Repository) GetByEmail(ctx context.Context, email string) (entity.User,
 	return u, nil
 }
 
+func (r *Repository) UserExists(ctx context.Context, email string) bool {
```
Comment (igkostyuk) [inline internal/repository/postgres/user/repo.go:117]: you can take user by email, and check error

Comment (glbter) [reply internal/repository/postgres/user/repo.go:117]: @igkostyuk do we need to take a user from a db, or our user would be lightwight as it is now?
As this aproach uses the advantages of using SQL queries over ORMs

File: internal/server/http/user/handlers.go (hunk 1 above)
Comment (glbter) [reply internal/server/http/user/handlers.go]: @igkostyuk I will think about it. If we are going to use it as other service, I will move it out soon.

**Comment by glbter** [comment]
@glbter need to fix change password: forgot to save new hash

File: internal/entity/user.go [hunk 3]
```diff
@@ -13,9 +13,19 @@ type User struct {
...
-	Name            string `json:"name,omitempty" validate:"omitempty"`
-	Email           string `json:"email" validate:"required,email"`
+	ID    string `json:"id" validate:"omitempty"`
+	Name  string `json:"name,omitempty" validate:"omitempty"`
+	Email string `json:"email" validate:"required,email"`
+	PasswordsRequest
+}
+
+type PasswordsRequest struct {
+	UserID          string `json:"id" validate:"omitempty"`
```
Comment (igkostyuk) [inline internal/entity/user.go:23]: in user with ID  userID :smile: 
I need to do something with user entity, right now it's too much inside it, maybe i will move it to handlers.

File: internal/repository/postgres/user/repo.go (hunk 2 above)
Comment (igkostyuk) [reply internal/repository/postgres/user/repo.go:117]: lightweight user ? do you worry about performance issues? do we need user exist by id ?

File: internal/server/http/user/handlers.go (hunk 1 above)
Comment (igkostyuk) [reply internal/server/http/user/handlers.go]: for this change user handlers can't exist without paswCases

File: internal/repository/postgres/user/repo.go (hunk 2 above)
Comment (igkostyuk) [reply internal/repository/postgres/user/repo.go:117]: do we need userExist by id ? do we feel any performance issues ? lightwight user ?

File: internal/entity/user.go (hunk 3 above)
Comment (glbter) [reply internal/entity/user.go:23]: Heh, actually, I need only id from user there. I did not embed it.
I thought about Roman's idea. Possible solution is to store passwords separately

Comment (igkostyuk) [reply internal/entity/user.go:23]: not sure how to do it right, need some research and rewrite it couple of times and test

Comment (igkostyuk) [reply internal/entity/user.go:23]: I will do something like this https://threedots.tech/post/common-anti-patterns-in-go-web-applications/

File: internal/server/http/handlers/user/handlers.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
+package user
```
Comment (igkostyuk) [inline internal/server/http/handlers/user/handlers.go:1]: can you explain why do delet my file and replace it with yours ?

**Review by igkostyuk — CHANGES_REQUESTED** [review]

File: internal/server/http/handlers/user/handlers.go (hunk 4 above)
Comment (glbter) [reply internal/server/http/handlers/user/handlers.go:1]: @igkostyuk I did not. that is the thing. Just moved it to a descriptive package - handlers

File: internal/entity/user.go (hunk 3 above)
Comment (glbter) [reply internal/entity/user.go:23]: @igkostyuk As I understood, your idea is to use less embedding but to make several entities. Or you talk about dto and entities?

File: internal/server/http/handlers/user/handlers.go (hunk 4 above)
Comment (igkostyuk) [reply internal/server/http/handlers/user/handlers.go:1]: yes you are. I lose all history.

Comment (glbter) [reply internal/server/http/handlers/user/handlers.go:1]: ok, if it is a trouble for u, we will leave this file out

Comment (igkostyuk) [reply internal/server/http/handlers/user/handlers.go:1]: it's not only for me

Comment (glbter) [reply internal/server/http/handlers/user/handlers.go:1]: I don't know what are you going to achive... It is ok for you to leave it where you created it?

Comment (igkostyuk) [reply internal/server/http/handlers/user/handlers.go:1]: you just delete 192 lines without a reason, i will show you in monday. you can check diff

Comment (glbter) [reply internal/server/http/handlers/user/handlers.go:1]: @igkostyuk Я правда не понимаю к чему спор.  Причину я обьяснил. Я не против оставить как было и сделать по твоему.

Comment (igkostyuk) [reply internal/server/http/handlers/user/handlers.go:1]: не удаляй файлы двигай их. в истории непонятно что ты там менял как ты менял.. ты все мои хендлеры удалил и заменил их на свои.

**Pull request closed without merging**